modules/fvad/fvad.c

```diff
@@ -0,0 +1,360 @@
+		return EINVAL;
+	}
+
+	st = mem_zalloc(sizeof(*st), enc_destructor);
+	if (!st)
+		return ENOMEM;
+
+	st->fvad = fvad_new();
+	if (!st->fvad) {
+		mem_deref(st);
+		return ENOMEM;
+	}
+
+	int err = fvad_set_sample_rate(st->fvad, prm->srate);
+	if (err < 0) {
+		warning("fvad: sample rate %d is not supported\n",
+			prm->srate);
+		mem_deref(st);
+		return EINVAL;
+	}
+
+	if (!st->call) {
+		struct filter_arg fa = { au, NULL };
+
+		uag_filter_calls(find_first_call, find_call, &fa);
+		st->call = fa.call;
+	}
+
+	*stp = (struct aufilt_enc_st *)st;
+
+	return 0;
+}
+
+
+static int decode_update(struct aufilt_dec_st **stp, void **ctx,
+			 const struct aufilt *af, struct aufilt_prm *prm,
+			 const struct audio *au)
+{
+	struct vad_dec *st;
+	(void)ctx;
+
+	if (!stp || !af || !prm)
+		return EINVAL;
+
+	if (*stp)
+		return 0;
+
+	if (prm->ch != 1) {
+		warning("fvad: only mono is supported\n");
+		return EINVAL;
+	}
+
+	if (prm->fmt != AUFMT_S16LE) {
+		warning("fvad: only AUFMT_S16LE is supported. "
+			"Use the auconv module to fix this\n");
+		return EINVAL;
+	}
+
+	st = mem_zalloc(sizeof(*st), dec_destructor);
+	if (!st)
+		return ENOMEM;
+
+	st->fvad = fvad_new();
+	if (!st->fvad) {
+		mem_deref(st);
+		return ENOMEM;
+	}
+
+	int err = fvad_set_sample_rate(st->fvad, prm->srate);
+	if (err < 0) {
+		warning("fvad: sample rate %d is not supported\n",
+			prm->srate);
+		mem_deref(st);
+		return EINVAL;
+	}
+
+	if (!st->call) {
+		struct filter_arg fa = { au, NULL };
+
+		uag_filter_calls(find_first_call, find_call, &fa);
+		st->call = fa.call;
+	}
+
+	*stp = (struct aufilt_dec_st *)st;
+
+	return 0;
+}
+
+
+static bool auframe_vad(Fvad *fvad, struct auframe *af)
+{
+	static int chunk_times_ms[] = { 30, 20, 10 };
+
+	if (af->fmt != AUFMT_S16LE) {
+		warning("fvad: invalid sample format %d\n",
+			af->fmt);
+
+		return false;
+	}
+
+	size_t pos = 0;
+
+	/* process all chunk_sizes that fvad accepts */
+	for (size_t chunk_time_index = 0;
+		chunk_time_index < RE_ARRAY_SIZE(chunk_times_ms);
+		++chunk_time_index) {
+
+		const size_t chunk_time = chunk_times_ms[chunk_time_index];
+		size_t sampc = af->srate / 1000 * chunk_time;
+
+		while (af->sampc - pos >= sampc) {
+
+			int err = fvad_process(fvad, (int16_t*)af->sampv + pos,
+				sampc);
+			pos += sampc;
+			if (err > 0) {
+				return true;
+			}
+			else if (err < 0) {
+				warning("fvad: fvad_process(%d) failed\n",
+					sampc);
+				return false;
+			}
+		}
+	}
+
+	if (pos != af->sampc) {
+		warning("fvad: fvad_process: samples left over: %d\n",
+			af->sampc - pos);
+	}
+
+	return false;
+}
+
+
+static int encode(struct aufilt_enc_st *st, struct auframe *af)
+{
+	struct vad_enc *vad = (void *)st;
+
+	if (!st || !af)
+		return EINVAL;
+
+	bool vad_tx = auframe_vad(vad->fvad, af);
+
+	if (vad_tx != vad->vad_tx) {
+		vad->vad_tx = vad_tx;
+
+		if (vad_stderr)
+			print_vad(61, 32, false, vad_tx);
+
+		module_event("fvad", "vad", call_get_ua(vad->call), vad->call,
+			"%d", vad_tx);
```

Comment:
Please align under `"fvad"`

Reply:
What do you mean with _align_ here? Rename `print_vad` to `print_fvad`?

Reply:
It makes sense to distinguish between rx ad tx, so I intend to do:
```
module_event("fvad", "vad_tx", call_get_ua(vad->call), vad->call, vad_tx ? "on" : "off");
```

(and `rx` for the decoding side)

Reply:
Align:
```c
		module_event("fvad", "vad_tx", call_get_ua(vad->call), vad->call,
			     vad_tx ? "on" : "off");
```